sycl/doc/KernelProgramCache.md

```diff
@@ -0,0 +1,82 @@
+# A brief overview of kernel/program caching mechanism.
+
+The cache is employed when one submits kernel for execution or builds program or
```

Comment:
I find it quite hard to follow the narration, frankly. Can we somewhat restruct the text?
Intro
  - what caches SYCL runtime implements (what is cached)
  - why (performance? why low performance w/o caches)
  - cache key and cache value (conceptually, w/o synchronization details at this point)
  - when caching happens, when cached value is used
  - mention as TODO that there is no eviction (https://github.com/intel/llvm/issues/2517)
  - why synchronization is needed
Details
  - details on when caching is skipped, when it is done.
  - overall synchronization scheme - what resource is guarded, what algorithm and primitives are used,
    what is sync primitive-to-resource correspondence (1-1, 1-many).
 - would be good to gave a picture with threads, cache elements and sync primitives

Reply:
I believe the readability of the document is improved now.